### PR TITLE
Allow BuildAgainstPackages to work with global test project.json 

### PR DESCRIPTION
**Pull request by karajas**
Check for use of common test project.json

Change AddDependenciesToProjectJson to allow for compile only assets in framework specific dependencies of common project.json

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 1]
```diff
@@ -10,6 +10,7 @@
 using NuGet.Packaging;
 using NuGet.Packaging.Core;
 using NuGet.Versioning;
+using System.Diagnostics;
```
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: will remove.

**Comment by karajas** [comment]
@chcosta

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets [hunk 2]
```diff
@@ -38,7 +38,7 @@
     </ResolveNuGetPackagesDependsOn>
   </PropertyGroup>
 
-  <Target Name="GenerateTestProjectJson" DependsOnTargets="ExtractProjectReferences;GatherProjectReferenceForProjectJson;DetermineProjectJsonPath;AddDependenciesToProjectJson" />
+  <Target Name="GenerateTestProjectJson" DependsOnTargets="ExtractProjectReferences;GatherProjectReferenceForProjectJson;DetermineProjectJsonPath;AddDependenciesToProjectJson" Condition="'$(UsesCommonTestProjectJson)'!='true'"/>
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: Where does UsesCommonTestProjectJson get set?

Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: By the way regardless of the condition value all the DependsOnTargets will still get executed, is that going to cause issues?

Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: I initially had the check on the depends on targets but this seems to have the same behavior.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets [hunk 3]
```diff
@@ -38,11 +38,11 @@
     </ResolveNuGetPackagesDependsOn>
   </PropertyGroup>
 
-  <Target Name="GenerateTestProjectJson" DependsOnTargets="ExtractProjectReferences;GatherProjectReferenceForProjectJson;DetermineProjectJsonPath;AddDependenciesToProjectJson" />
+  <Target Name="GenerateTestProjectJson" DependsOnTargets="ExtractProjectReferences;GatherProjectReferenceForProjectJson;DetermineProjectJsonPath;AddDependenciesToProjectJson"/>
 
   <Target Name="GenerateAllTestProjectJsons" DependsOnTargets="GenerateTestProjectJson" />
 
-  <Target Name="DetermineProjectJsonPath">
+  <Target Name="DetermineProjectJsonPath" Condition="'$(UsesCommonTestProjectJson)'!='true'">
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: I definitely don't like seeing this condition duplicated all over the place. You should use something along the lines of 

```
  <PropertyGroup>
    <GenerateTestProjectJsonDependsOn Condition="'$(SkipGenerateTestProjectJson)'!='true'">
      ExtractProjectReferences;
      GatherProjectReferenceForProjectJson;
      DetermineProjectJsonPath;
      AddDependenciesToProjectJson
    </GenerateTestProjectJsonDependsOn>
  </PropertyGroup>
 
  <Target Name="GenerateTestProjectJson" DependsOnTargets="$(GenerateTestProjectJsonDependsOn)" />
```

Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: Made the associated change. Since the propertygroup listed above is global, I had to create another target to evaluate in the given context.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets [hunk 4]
```diff
@@ -37,8 +37,21 @@
       DetermineProjectJsonPath;
     </ResolveNuGetPackagesDependsOn>
   </PropertyGroup>
-
-  <Target Name="GenerateTestProjectJson" DependsOnTargets="ExtractProjectReferences;GatherProjectReferenceForProjectJson;DetermineProjectJsonPath;AddDependenciesToProjectJson" />
+  
+  <!--This target is required to evaluate this property uniquely in the context of each project -->
+  <Target Name="EvaluateGenerateTestProjectJsonDependsOn" >
+    <PropertyGroup>
+      <GenerateTestProjectJsonDependsOn Condition="'$(SkipGenerateTestProjectJson)'=='false'">
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: This shouldn't be in a wrapper target. Also I would change this condition to `'$(SkipGenerateTestProjectJson)' == 'true'`

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets [hunk 5]
```diff
@@ -38,8 +38,6 @@
     </ResolveNuGetPackagesDependsOn>
   </PropertyGroup>
   
-  <!--This target is required to evaluate this property uniquely in the context of each project -->
-  <Target Name="EvaluateGenerateTestProjectJsonDependsOn" >
     <PropertyGroup>
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/buildagainstpackages.targets]: nit: indention should align.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 6]
```diff
@@ -275,13 +275,13 @@ private static JObject ReadProject(string projectJsonPath)
...
-        private JToken GetFrameworkDependenciesSection(JObject projectJsonRoot, string framework = null)
+        private JObject GetFrameworkDependenciesSection(JObject projectJsonRoot, string framework = null)
         {
             if(string.IsNullOrWhiteSpace(framework))
             {
-                return projectJsonRoot["dependencies"];
+                return (JObject) projectJsonRoot["dependencies"];
             }
-            return projectJsonRoot.SelectToken("frameworks." + NewtonsoftEscapeJProperty(framework) + ".dependencies");
+            return (JObject) projectJsonRoot["frameworks"][framework]["dependencies"];
```
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: The other two occurrences look like they want to determine the path to the entry, and so use the NewtonsoftEscapeJProperty method to construct it and refer to it as a JProperty. Here I wanted to retrieve the JObject that is referenced at this path.

**Comment by chcosta** [comment]
@karajas, can you clarify for me why adddependenciestoprojectjson is still needed when using a common project.json?

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs (hunk 6 above)
Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: I don't see anything about the other two instances which necessitates they be treated as JProperties.

**Comment by chcosta** [comment]
Are there test projects with project references to things not in the common project.json?

**Comment by karajas** [comment]
Yup there are still test project.json's that are local to the test project.
Also adddependenciestoprojectjson is still used to update dependencies on a project.json of a locally build project, the VersionTools repo update script had some edge cases that were more easily resolve with AddDependencies.
